scripts/zkai_stwo_query_grinding_budget_gate.py

```
@@ -0,0 +1,555 @@
+        raise StwoQueryGrindingBudgetGateError(f"inventory field {field} is invalid") from err
+
+
+def load_inventory() -> list[dict[str, Any]]:
+    raw = hook_gate.read_repo_file(INVENTORY_TSV, "predecommit inventory TSV", hook_gate.MAX_EVIDENCE_BYTES)
+    try:
+        text = raw.decode("utf-8")
+    except UnicodeDecodeError as err:
+        raise StwoQueryGrindingBudgetGateError("predecommit inventory TSV is not UTF-8") from err
+    rows = list(csv.DictReader(io.StringIO(text), delimiter="\t"))
+    if len(rows) != EXPECTED_INVENTORY_COUNT:
+        raise StwoQueryGrindingBudgetGateError("inventory row count drift")
```

Ruff (0.15.13)

[warning] Avoid specifying long messages outside the exception class

(TRY003)